### PR TITLE
Add Joint Distribution

### LINKED ISSUE #21: Add Joint Distribution
The next goal is to add a joint distribution class that can combine multiple distributions into a single object. This is how the user would specify a "Bayesian model" for an inverse problem.

Initially it would be worthwhile simply to add the Joint Distribution class as discussed in previous issues on Gitlab. The main goal of the joint distribution class would be to allow a simple interface for conditioning and evaluating the log density function for multiple parameters.

The dimension and geometry of the joint distribution should be a list of ints and Geometry respectively.

### PR DESCRIPTION
**Pull request by nabriis**
Closes #21 

- [x] Add Joint distribution
- [x] Add Docs
- [x] Add tests
- [x] Add Stacked Joint Distribution
- [x] Add tests for stacked
- [x] Add  tutorial

The tutorial file file can be compiled to ipynb same way we compile docs
```
sphinx-build -b html docs/. docs/_build
```
and overview with links will be under _build\user\_auto_tutorials\JointDistribution.html in cuqipy directory

### CODE REVIEW

**Comment by nabriis** [comment]
Hi @amal-ghamdi. I added you as initial reviewer on this PR.

**Review by jakobsj — COMMENTED** [review]
Being a distribution, should joint distribution have a name?  At the moment `joint.name` fails since it does not a have a name.  This relates to the discussion of whether "name" or "random variable" etc should be used. It is not clear how one might define a "name" for joint based on the names of the densities "x" and "y" it is constructed from, but it seems okay to let it have a list of random variables ["x", "y"] similarly to geometry and dim.

File: demos/tutorials/JointDistribution.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+# %%
+# Basic properties of the joint distribution
+# ------------------------------------------
+#
+# The joint distribution comes with two main properties:
+# dim (dimension) and geometry. Because its a joint distribution
+# over multiple variables, these are lists.
+
+print(joint.dim)
```
Comment (jakobsj) [inline demos/tutorials/JointDistribution.py:72]: The dimension of y is None, is that intended? Being defined through a lambda function its size is perhaps hard to know in general?

File: demos/tutorials/JointDistribution.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+# Define fixed parameters
+d = 0.1
+s = 0.05
+
+# Define distributions
+x = GaussianCov(np.zeros(n), d*In)
+y = GaussianCov(lambda x: A@x, s*Im)
+
+# Define joint distribution p(x,y)
+joint = JointDistribution([x, y])
```
Comment (jakobsj) [inline demos/tutorials/JointDistribution.py:59]: We construct the joint from the two parameters x and y. But `joint.get_parameter_names()` returns the four parameters
`['d', 'l', 'x', 'y']`  and also prints the four parameter
```

print(joint)
JointDistribution(
    Equation: 
	p(d,l,x,y) = p(d)p(l)p(x|d)p(y|x,l)
    Densities: 
	d ~ CUQI Gamma.
	l ~ CUQI Gamma.
	x ~ CUQI GaussianCov. Conditioning variables ['d'].
	y ~ CUQI GaussianCov. Conditioning variables ['x', 'l'].
)
```
That is not completely intuitive?  We don't need to define the joint from all four?

File: demos/tutorials/JointDistribution.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+# Density evaluation
+# ------------------
+#
+# A main method of the joint distribution is to be able to evaluate the
+# un-normalized log density function (logd) given a set of values for the
+# variables. This can be achieved by passing either positional or keyword
+# arguments.
+
+# Using keyword arguments
+logd = joint.logd(y=np.ones(m), x=np.ones(n))
```
Comment (jakobsj) [inline demos/tutorials/JointDistribution.py:85]: Related to the above point on two vs four parameters, here values of the two which completely specifies everything so we can evaluate the logd. However, if we, for whatever reason maybe as mistake, specify just one, e.g., 
```
joint.logd(y=np.ones(m))
```
we get an error that tells us we need to specify all four:
```
ValueError: To evaluate the log density function, all parameters must be passed. Received dict_keys(['y']) and expected ['d', 'l', 'x', 'y'].
```

File: demos/tutorials/JointDistribution.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+Cd._reduce_to_single_density()
+
+# %%
+# Another "hack" is to have the joint distribution act as if it were a single
+# parameter density. This is achieved by calling the `._as_stacked()` method.
+# This returns a new "stacked" joint distribution that the samplers/solvers
+# can use as if it were any other Density.
+
+posterior_stacked = joint(y=y_obs)._as_stacked()
```
Comment (jakobsj) [inline demos/tutorials/JointDistribution.py:225]:
```suggestion
posterior_stacked = posterior._as_stacked()
```

Comment (jakobsj) [reply demos/tutorials/JointDistribution.py:225]: Reusing already created `posterior` for clarity and simplicity.

File: demos/tutorials/JointDistribution.py (hunk 1 above)
Comment (jakobsj) [reply demos/tutorials/JointDistribution.py:72]: I think we agreed you would create an issue on allowing the user to specify dimension manually?

**Review by jakobsj — CHANGES_REQUESTED** [review]
I went through the demo again and that resolved the issues about two or four parameters, my fault.  Super nice work, thank you for adding the comprehensive demo and for the explanations earlier, that helped a lot in reviewing this. Only added some minor requests, and re `.name` of joints and posteriors and option to equip distributions with dimension manually please create issues. I haven't given much scrutiny to the private helper functions.

File: demos/tutorials/JointDistribution.py (hunk 1 above)
Comment (nabriis) [reply demos/tutorials/JointDistribution.py:72]: #28

File: demos/tutorials/JointDistribution.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+#
+# .. note::
+#
+#     The posterior distribution as shown above can be sampled via the sampler module.
+#     See e.g. the tutorial on sampling :doc:`Sampling tutorial <../_auto_tutorials/4-Samplers>`.
+#
+# Because the posterior no longer depends on the data, the dimension is only
+# with respect to the parameters.
+print(f"Posterior dimension: {posterior.dim}")
+print(f"Posterior geometry: {posterior.geometry}")
```
[the remark this answers is not in the document]
Comment (nabriis) [reply demos/tutorials/JointDistribution.py:125]: The Posterior class does have the option to get a name attribute similar to any other Distribution. But the "posterior" from the demo is still a joint distribution. Added an issue to rethink this name attribute #29

File: demos/tutorials/JointDistribution.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+# We can write this model in CUQIpy as follows:
+
+# Define distribution
+d = Gamma(1, 1e-4)
+l = Gamma(1, 1e-4)
+x = GaussianCov(np.zeros(n), lambda d: d*In)
+y = GaussianCov(lambda x: A@x, lambda l: l*Im)
+
+# Define joint distribution p(d,l,x,y)
+joint = JointDistribution([d, l, x, y])
```
[the remark this answers is not in the document]
Comment (nabriis) [reply demos/tutorials/JointDistribution.py]: Done. Thanks :)

**Comment by nabriis** [comment]
> Being a distribution, should joint distribution have a name? At the moment `joint.name` fails since it does not a have a name. This relates to the discussion of whether "name" or "random variable" etc should be used. It is not clear how one might define a "name" for joint based on the names of the densities "x" and "y" it is constructed from, but it seems okay to let it have a list of random variables ["x", "y"] similarly to geometry and dim.

Hi @jakobsj. Thank you for the review. I incorporated most of your comments. I added two new issues coming from our discussions as you can see in the comments. Finally after our retro I also split the demo file for the joint distribution into a tutorial relevant to users and a developer demo. I added the developer demo in the contributor guide under a new section.

**Review by jakobsj — APPROVED** [review]
Good idea to split off a technical demo and starting a collection of such demos.

File: demos/tutorials/JointDistribution.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+#
+# The joint distribution is a tool that allows us to define Bayesian models.
+# This is achieved by combining multiple distributions into a single object.
+#
+# For example, consider the following model
+#
+# .. math::
+#
+#   \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#   \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
```
Comment (amal-ghamdi) [inline demos/tutorials/JointDistribution.py:43]: same comment about d and s here (as in dev/JointDistribution.py)

File: demos/tutorials/JointDistribution.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+#
+# For example, consider the following extension of the Bayesian model earlier:
+#
+# .. math::
+#
+#     \begin{align}
+#         d &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         l &\sim \mathrm{Gamma}(1,10^{-4}) \\
+#         \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#         \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
```
Comment (amal-ghamdi) [inline demos/tutorials/JointDistribution.py:142]: same comment about d and s here (as in dev/JointDistribution.py)

**Review by amal-ghamdi — COMMENTED** [review]
Thank you @nabriis for adding these great features! The PR is very nice. I had few minor comments, mostly about documentation.

File: cuqi/distribution/_joint_distribution.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        if len(args)>0:
+            ordered_keys = self.get_parameter_names()
+            for index, arg in enumerate(args):
+                if ordered_keys[index] in kwargs:
+                    raise ValueError(f"{ordered_keys[index]} passed as both argument and keyword argument.\nArguments follow the listed parameter names order: {ordered_keys}")
+                kwargs[ordered_keys[index]] = arg
+        return kwargs
+
+    def _reduce_to_single_density(self):
+        """ Reduce the joint distribution to a single density if possible.
```
[the remark this answers is not in the document]
Comment (nabriis) [reply cuqi/distribution/_joint_distribution.py:183]: Good suggestion. Thanks

File: tests/test_joint_distribution.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    J = cuqi.distribution.JointDistribution(densities)
+
+    # Get the parameter names
+    names = J.get_parameter_names()
+
+    # Get dimensions
+    dims = J.dim
+
+    # Make dictionary of random parameters matching names and dimensions
+    params = {name: abs(np.random.randn(dim)) for name, dim in zip(names, dims)}
```
[the remark this answers is not in the document]
Comment (nabriis) [reply tests/test_joint_distribution.py:76]: Hmm. The random number was part of the test. If it starts randomly failing for no good reason we can consider adding a seed.

File: demos/dev/JointDistribution.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+#
+# For example, consider the following extension of the Bayesian model earlier:
+#
+# .. math::
+#
+#     \begin{align}
+#         d &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         l &\sim \mathrm{Gamma}(1,10^{-4}) \\
+#         \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#         \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
```
[the remark this answers is not in the document]
Comment (nabriis) [reply demos/dev/JointDistribution.py:42]: Good catch. Was a typo. I updated it to d^{-1} and l^{-1} to match code

File: demos/dev/JointDistribution.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+#
+# That is, we need to define
+#
+# .. math::
+#
+#       C_\mathbf{x}&=p(\mathbf{x} \mid \hat{\mathbf{y}}, \hat{d}, \hat{l})\\
+#       C_d&=p(d \mid \hat{\mathbf{y}}, \hat{\mathbf{x}}, \hat{l})\\
+#       C_l&=p(l \mid \hat{\mathbf{y}}, \hat{\mathbf{x}}, \hat{d})
+#
+# Assuming we have some values for :math:`\mathbf{x}`, :math:`d` and :math:`l`,
```
[the remark this answers is not in the document]
Comment (nabriis) [reply demos/dev/JointDistribution.py:79]: Added. Thanks

**Comment by nabriis** [comment]
Thanks for the review @amal-ghamdi. Good you caught those inconsistencies!

**Review by amal-ghamdi — APPROVED** [review]
Looks great! Thank you @nabriis.

**Pull request merged**